Small improvements to str_cat

hack-c opened the pull request:
closes #11435 
closes #11334.

Added some doc examples to make the behavior with `nan` more clear, and catch an error condition when a `sep` is passed instead of an `others` (array) as an argument and print a more intuitive error message.

Commit:
ENH DOC added some new doc examples to str_cat and catch an error to prompt more clearly for a sep keyword

Commit:
ENH DOC added some new doc examples to str_cat and catch an error to prompt more clearly for a sep keyword

Commit:
ENH DOC added some new doc examples to str_cat and catch an error to prompt more clearly for a sep keyword

Commit:
Merge branch 'str-cat-error' of https://github.com/hack-c/pandas into str-cat-error

jreback commented:
tests!

Commit:
TST added test to make sure improved error fires

Commit:
TST added test to make sure improved error fires

jreback commented on pandas/tests/test_strings.py:
```diff
@@ -2057,6 +2059,17 @@ def test_method_on_bytes(self):
             tm.assert_series_equal(result, expected)
 
+    def test_str_cat_raises_intuitive_error(self):
+        s = Series(['a','b','c','d'])
```
add the issue number as a comment here

jreback commented on pandas/core/strings.py:
```diff
@@ -45,6 +45,15 @@ def str_cat(arr, others=None, sep=None, na_rep=None):
+    When ``na_rep`` is `None` (default behavior), NaN value(s)
+    in the Series propagate and return value will be NaN.
+
+    >>> Series(['a','b',np.nan,'c']).str.cat(sep=' ')
```
I think better to actually ignore nans (as indicated in the issue), rather than just document

hack-c replied:
Do you want to just treat them as `''` ?

hack-c replied:
Or as if they weren't present in the original array?

i.e. do a fillna then cat, or dropna then cat?

@jreback

jreback replied:
I think ``.dropna`` (which doesn't exist on ``Index``) - its pretty easy (there is an issue about it).

so you can emulate so it will just work (for Series or Index)

``si.take(pd.notnull(si))``

hack-c replied:
But then you're changing the array length and what if someone supplies `others` w/ length equal to the original array (before `dropna`)...?

hack-c replied:
The `fillna` way amounts to changing the default value of `na_rep` to `''`.

hack-c replied:
Maybe in that bad case for `others` you just throw a `ValueError`.

jreback replied:
actually you bring up a good point. doing ``.fillna(na_rep)``, THEN processing might just work and do the right thing.

yes if ``others`` is then inconsistent you need to raise.

pls add as many test cases as you think are needed to represent all behavior.

jreback commented:
pls also add entries in whatsnew (one for each of the bug fixes). we'll call ignoring nans a bug fix.

jreback commented:
can you rebase / update

Commit:
add comment with issue #

Commit:
add comment with issue #

jreback committed:
Merge pull request #12111 from jreback/matrix

CLN: remove core/matrix.py, not imported into the pandas namespace

kawochen committed:
CLN: cleaned RangeIndex._min_fitting_element

Added test cases for `_min_fitting_element` and `_max_fitting_element`
(unused) that would fail in master because of precision

Author: Ka Wo Chen <kawoc@tepper.cmu.edu>

Closes #12113 from kawochen/CLN-RI-max-fitting and squashes the following commits:

3cbc7f5 [Ka Wo Chen] CLN: cleaned RangeIndex._min_fitting_element

wesm committed:
CLN: fix flake8 warnings in pandas/stats

Author: Wes McKinney <wes@cloudera.com>

Closes #12114 from wesm/style/stats and squashes the following commits:

a3f8508 [Wes McKinney] CLN: fix flake8 warnings in pandas/stats

wesm committed:
CLN: grab bag of flake8 fixes

Author: Wes McKinney <wes@cloudera.com>

Closes #12115 from wesm/style/flake8-misc and squashes the following commits:

017ca16 [Wes McKinney] CLN: grab bag of flake8 fixes

Commit:
ENH changed behavior of str_cat to drop NaNs rather than returning NaN in case of array with NaN.

Commit:
merge conflict

hack-c commented:
I may have just botched this... tried rebasing on master and now there are many changes I didn't include reflected in the diff.

jreback commented:
http://pandas.pydata.org/pandas-docs/stable/contributing.html#contributing-your-changes-to-pandas

``git rebase -i master``

then

``git push <yourremote> <yourbranch> -f``

wesm commented:
You have to make sure your *your* master is in sync with upstream master. 

You can revert to your original patch with "git reset --hard HASH" -- look in your git reflog to find the commit hash.

hack-c commented:
This was botched and left to languish-- have restarted and will ref this in the re-do.
